library/include/BitArray.h

```diff
@@ -232,4 +234,299 @@ namespace DFHack
+            }
+            bool operator!=(const iterator & other) const
+            {
+                return !(*this == other);
+            }
+        };
+        class const_iterator
+        {
+            iterator iter;
+            friend struct DfLinkedList<L, I>;
+        public:
+            using difference_type = void;
+            using value_type = I * ;
+            using pointer = I * const *;
+            using reference = const proxy;
+            using iterator_category = std::bidirectional_iterator_tag;
+
+            const_iterator(const iterator & iter) : iter(iter) {}
+            const_iterator(const const_iterator & other) : iter(other.iter) {}
+
+            const_iterator & operator++()
+            {
+                ++iter;
+                return *this;
+            }
+            const_iterator & operator--()
+            {
+                --iter;
+                return *this;
+            }
+            const_iterator operator++(int)
+            {
+                const_iterator copy(*this);
+                ++*this;
+                return copy;
+            }
+            const_iterator operator--(int)
+            {
+                const_iterator copy(*this);
+                --*this;
+                return copy;
+            }
+            const_iterator & operator=(const const_iterator & other)
+            {
+                iter = other.iter;
+                return *this;
+            }
+            const proxy operator*() const
+            {
+                return *iter;
+            }
+            I *operator->() const
+            {
+                return *iter;
+            }
+            bool operator==(const const_iterator & other) const
+            {
+                return iter == other.iter;
+            }
+            bool operator!=(const const_iterator & other) const
+            {
+                return iter != other.iter;
+            }
+        };
+
+        typedef I *value_type;
+        typedef I * & reference_type;
+        typedef void difference_type;
+        typedef size_t size_type;
+
+        bool empty() const
+        {
+            return static_cast<const L *>(this)->next == nullptr;
+        }
+        size_t size() const
+        {
+            size_t n = 0;
+            for (I *i : *this)
+                n++;
+            return n;
+        }
+
+        iterator begin()
+        {
+            return iterator(static_cast<L *>(this), static_cast<L *>(this));
+        }
+        const_iterator begin() const
+        {
+            return const_iterator(const_cast<DfLinkedList<L, I> *>(this)->begin());
+        }
+        const_iterator cbegin() const
+        {
+            return begin();
+        }
+        iterator end()
+        {
+            L *it = static_cast<L *>(this);
+            while (it->next && it->next->next)
+            {
+                it = it->next;
+            }
+            return iterator(static_cast<L *>(this), it, true);
+        }
```

Comment:
end should be as simple as return iterator(nullptr) because the end should point to element after the last valid element. But that wouldn't be perfectly valid because std::prev(end()) is required to work.

Reply:
For forward iterators? I don't think we have a fast way to go backwards (or start at the end of a list) here.

Reply:
dflist is double linked list expect its root isn't correctly linked to end and begin. But if we would provide only forward iterator and simple implement `begin() { return iterator(root->next)}` and `end() {return iterator(nullptr);}` and adapt iterator to simple keep pointer to df::list_link. Then end iterator would compare equal only when iterator has been increment one past the end.

Reply:
Oh, right, I thought it was singly-linked (probably because Ben was implementing a forward iterator).

Reply:
For practical use it is same as singly-linked list except it makes erase a bit simpler. But the link between root and end would need to be added to have all double-linked list features.

Reply:
Is `std::prev` required to have a specific time complexity? I could make the end of the list be a special case where it computes it on demand.

Reply:
complexity requirements and other are mostly only for stl implementations. DF list can't be made easily conform to std::list requirements. The hardest requirement is that size() must be O(1). But std::forward_list requirements are something that DF list can fairly easily match. There it would have only forward iterator which would mean that std::prev(end()) isn't valid use. But then we could mix that with std::list interface for simpler usage as we actually have double-linked list which makes it easy to implement erase and insert with only current list_link pointed by iterator.
http://en.cppreference.com/w/cpp/container/forward_list
http://en.cppreference.com/w/cpp/container/list

I think that would then allow dfhack code use our "std::list" like container with most important stuff like range-for loop, stl algorithms and std::list like erase and insert interface.

Reply:
Actually. There is the performance issue of calling list::insert(end(), new_entry). But we could disallow use of that kind of construct.

Reply:
I've changed it so the only performance drawback on `end()` will be when moving it. You can compare with it for basically free now.